Naive CPU kernels for jagged <-> padded dense conversions

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121947
* #126552
* #125947
* #122836
* #128057
* __->__ #127007

This PR introduces naive CPU impls for:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`

On the CUDA side, these are backed by lifted FBGEMM kernels. We may want to revisit the CPU versions with higher-performance implementations at a later time.

jbschlosser committed:
Naive CPU kernels for jagged <-> padded dense conversions

[ghstack-poisoned]

jbschlosser committed:
Update on "Naive CPU kernels for jagged <-> padded dense conversions"




[ghstack-poisoned]

davidberard98 commented on aten/src/ATen/native/nested/NestedTensorTransformerFunctions.cpp:
```diff
@@ -246,5 +246,73 @@ Tensor NestedTensor_to_mask(const Tensor& nt, std::optional<int64_t> mask_dim, c
+      "_jagged_to_padded_dense_forward(): only a single jagged dim is supported for now");
+
+  // allocate appropriately-sized padded tensor
+  auto offsets = offsets_list[0];
```
nit: can we assert offsets is 1D?

jbschlosser replied:
yep good call! adding

davidberard98 commented on aten/src/ATen/native/nested/NestedTensorTransformerFunctions.cpp:
```diff
@@ -246,5 +246,73 @@ Tensor NestedTensor_to_mask(const Tensor& nt, std::optional<int64_t> mask_dim, c
+  // copy data to padded tensor
+  for (auto i : c10::irange(batch_size)) {
+    auto start_offset = offsets[i].item<int64_t>();
+    auto end_offset = offsets[i + 1].item<int64_t>();
```
I think we should set `end_offset = min(offsets[i+1].item<int64_t>(), start_offset + max_length)` in case any sequence is larger than max_length. Or if we don't want to support it right now, add an assert

davidberard98 commented on aten/src/ATen/native/nested/NestedTensorTransformerFunctions.cpp:
```diff
@@ -246,5 +246,73 @@ Tensor NestedTensor_to_mask(const Tensor& nt, std::optional<int64_t> mask_dim, c
+      "_jagged_to_padded_dense_forward(): only a single jagged dim is supported for now");
+
+  // allocate appropriately-sized values tensor
+  auto offsets = offsets_list[0];
```
nit: same here, assert offsets is 1D

davidberard98 replied:
also, maybe assert that `padded_shape.size() >= 2`?

jbschlosser replied:
Added check with test

davidberard98 commented on aten/src/ATen/native/nested/NestedTensorTransformerFunctions.cpp:
```diff
@@ -246,5 +246,73 @@ Tensor NestedTensor_to_mask(const Tensor& nt, std::optional<int64_t> mask_dim, c
+  auto batch_size = offsets.size(0) - 1;
+  for (auto i : c10::irange(batch_size)) {
+    auto start_offset = offsets[i].item<int64_t>();
+    auto end_offset = offsets[i + 1].item<int64_t>();
```
same here, I think if we have padded_shape[1] < length, we need to handle it here (and maybe pad with zeros for now?)... or assert otherwise

jbschlosser replied:
Added check with test

davidberard98 reviewed (commented):
nice! generally LGTM but could you add a few more asserts, and maybe some handling for sequences longer than max_length?

jbschlosser committed:
Update on "Naive CPU kernels for jagged <-> padded dense conversions"


This PR introduces naive CPU impls for:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`

On the CUDA side, these are backed by lifted FBGEMM kernels. We may want to revisit the CPU versions with higher-performance implementations at a later time.

[ghstack-poisoned]

jbschlosser commented on aten/src/ATen/native/nested/NestedTensorTransformerFunctions.cpp:
```diff
@@ -246,5 +246,73 @@ Tensor NestedTensor_to_mask(const Tensor& nt, std::optional<int64_t> mask_dim, c
   return result;
 }
 
+Tensor _jagged_to_padded_dense_forward_cpu(
```
good Q! `torchgen` will indeed utilize the info from `native_functions.yaml` to generate an op-specific header at `build/aten/src/ATen/ops/_jagged_to_padded_dense_forward_native.h`. The declarations for both CPU and CUDA impls are there:
```cpp
TORCH_API at::Tensor _jagged_to_padded_dense_forward_cpu(const at::Tensor & values, at::TensorList offsets, at::IntArrayRef max_lengths, double padding_value=0.0);
TORCH_API at::Tensor _fbgemm_jagged_to_padded_dense_forward(const at::Tensor & values, at::TensorList offsets, at::IntArrayRef max_lengths, double padding_value=0.0);
```

and the corresponding definitions are in the .cpp / .cu files.

jbschlosser committed:
Update on "Naive CPU kernels for jagged <-> padded dense conversions"


This PR introduces naive CPU impls for:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`

On the CUDA side, these are backed by lifted FBGEMM kernels. We may want to revisit the CPU versions with higher-performance implementations at a later time.

[ghstack-poisoned]

jbschlosser committed:
Update on "Naive CPU kernels for jagged <-> padded dense conversions"


This PR introduces naive CPU impls for:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`

On the CUDA side, these are backed by lifted FBGEMM kernels. We may want to revisit the CPU versions with higher-performance implementations at a later time.

[ghstack-poisoned]

jbschlosser committed:
Update on "Naive CPU kernels for jagged <-> padded dense conversions"


This PR introduces naive CPU impls for:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`

On the CUDA side, these are backed by lifted FBGEMM kernels. We may want to revisit the CPU versions with higher-performance implementations at a later time.

[ghstack-poisoned]

jbschlosser committed:
Update on "Naive CPU kernels for jagged <-> padded dense conversions"


This PR introduces naive CPU impls for:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`

On the CUDA side, these are backed by lifted FBGEMM kernels. We may want to revisit the CPU versions with higher-performance implementations at a later time.

[ghstack-poisoned]

davidberard98 reviewed (approved):
awesome, LGTM!

jbschlosser committed:
Update on "Naive CPU kernels for jagged <-> padded dense conversions"


This PR introduces naive CPU impls for:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`

On the CUDA side, these are backed by lifted FBGEMM kernels. We may want to revisit the CPU versions with higher-performance implementations at a later time.

[ghstack-poisoned]

jbschlosser committed:
Update on "Naive CPU kernels for jagged <-> padded dense conversions"


This PR introduces naive CPU impls for:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`

On the CUDA side, these are backed by lifted FBGEMM kernels. We may want to revisit the CPU versions with higher-performance implementations at a later time.

[ghstack-poisoned]

jbschlosser committed:
Update on "Naive CPU kernels for jagged <-> padded dense conversions"


This PR introduces naive CPU impls for:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`

On the CUDA side, these are backed by lifted FBGEMM kernels. We may want to revisit the CPU versions with higher-performance implementations at a later time.

[ghstack-poisoned]

jbschlosser committed:
Update on "Naive CPU kernels for jagged <-> padded dense conversions"


This PR introduces naive CPU impls for:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`

On the CUDA side, these are backed by lifted FBGEMM kernels. We may want to revisit the CPU versions with higher-performance implementations at a later time.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot merge

jbschlosser committed:
Update on "Naive CPU kernels for jagged <-> padded dense conversions"


This PR introduces naive CPU impls for:
* `_jagged_to_padded_dense_forward()`
* `_padded_dense_to_jagged_forward()`

On the CUDA side, these are backed by lifted FBGEMM kernels. We may want to revisit the CPU versions with higher-performance implementations at a later time.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
